File: ESAMI/_Svolti [OLD]/02-11-2017/java-aux/Queue.java
public class Queue <T>{


    private queueItem<T> head;
    private queueItem<T> tail;
    private int size;
    
    // ListaVuota
    public Queue() {
        head = null;
        tail = null;
        size = 0;
    }

    public boolean isEmptyQueue() {
        return size==0;
    }

    public void enqueue (T item){
    	queueItem<T> it = new queueItem<T>();
    	it.item=item;
    	it.next=null;
    	if (tail == null){ //&& primo ==null
    		tail=it;
    		head=it;
    	}else{
    		tail.next=it;
    		tail=it;
    	}
    	size++;
    }
    
    public T dequeue(){
    	if (isEmptyQueue()) return null;

    	queueItem<T> hd = head;
    	if (size==1){
    		head = null;
    		tail = null;
    	}else{
        	head = head.next;
    	}
    	size--;
    	return hd.item;
    }
    
    public int queueSize(){
    	return size;
    }
   
}

File: ESAMI/DOCS_NOTES/Fondamenti di Informatica II/AlgoritmiEStruttureDati-master/AlgoritmiEStruttureDati-master/Algoritmi - Esami/ALGORITMI/Esami/2017/13Lug/c-aux/queue.c
/* versione C */

/*
    realizza coda generica
*/

#ifndef QUEUE
#define QUEUE

#include <stdlib.h>
#include <assert.h>
#include "queue.h"

queue *newQueue() {
    queue *q = malloc(sizeof(queue));
    assert(q!=NULL);
    q->head=q->tail=NULL;
    q->size = 0;
    return q;
}

void enqueue(queue *q, void *item) {
    queueItem *it = malloc(sizeof(queueItem));
    assert(it!=NULL);
    it->item = item;
    it->next = NULL;
    if(q->tail != NULL) q->tail->next = it;
    q->tail = it;
    if(q->head == NULL) q->head = it;
    q->size++;
    return;
}

void *dequeue(queue *q) {
    queueItem *hd = q->head;
    if(hd==NULL) return NULL;
    void *ret;
    q->head = q->head->next; // se 1 solo elem. head diventa NULL
    q->size--;
    if(q->size==0) q->tail = NULL; // head è già a NULL
    /*
        equivalentemente:
        if(q->head==NULL) q->tail==NULL;
    */
    ret = hd->item;
    free(hd);
    return ret;
}

boolean isEmptyQueue(queue *q) {
    assert(q!=NULL);
    assert(q->size >= 0);
    if(q->size == 0) return TRUE;
    else return FALSE;
}

#endif
